base/compiler/typeinfer.jl

```diff
@@ -8,7 +8,192 @@ function typeinf(interp::AbstractInterpreter, result::InferenceResult, cached::B
     return typeinf(interp, frame)
 end
 
+"""
+The module `Core.Compiler.Timings` provides a simple implementation of nested timers that
+can be used to measure the exclusive time spent inferring each method instance that is
+recursively inferred during type inference.
+
+This is meant to be internal to the compiler, and makes some specific assumptions about
+being used for this purpose alone.
+"""
+module Timings
+
+using Core.Compiler: -, +, :, Vector, length, first, empty!, push!, pop!, @inline,
+    @inbounds, copy
+
+# What we record for any given frame we infer during type inference.
+function _typeinf_identifier(frame::Core.Compiler.InferenceState)
+    mi_info = (
+        frame.linfo,
+        frame.world,
+        copy(frame.sptypes),
+        frame.slottypes[2:end],  # Skip repeating the method instance
+    )
+    return mi_info
+end
+
+"""
+    Core.Compiler.Timing(mi_info, start_time, ...)
+
+Internal type containing the timing result for running type inference on a single
+MethodInstance.
+"""
+struct Timing
+    mi_info::Tuple{Core.MethodInstance,UInt64,Vector{Any},Vector{Any}}
+    start_time::UInt64
+    cur_start_time::UInt64
+    time::UInt64
+    children::Core.Array{Timing,1}
+    Timing(mi_info, start_time, cur_start_time, time, children) = new(mi_info, start_time, cur_start_time, time, children)
+    Timing(mi_info, start_time) = new(mi_info, start_time, start_time, UInt64(0), Timing[])
+end
+
+_time_ns() = ccall(:jl_hrtime, UInt64, ())  # Re-implemented here because Base not yet available.
+
+# We keep a stack of the Timings for each of the MethodInstances currently being timed.
+# Since type inference currently operates via a depth-first search (during abstract
+# evaluation), this vector operates like a call stack. The last node in _timings is the
+# node currently being inferred, and its parent is directly before it, etc.
+# Each Timing also contains its own vector for all of its children, so that the tree
+# call structure through type inference is recorded. (It's recorded as a tree, not a graph,
+# because we create a new node for duplicates.)
+const _timings = Timing[]
+# ROOT() is an empty function used as the top-level Timing node to measure all time spent
+# *not* in type inference during a given recording trace. It is used as a "dummy" node.
+function ROOT() end
+ROOT()  # Call it to compile a method instance for it.
```

Comment:
```suggestion
const ROOTmi = Core.Compiler.specialize_method(first(methods(ROOT)), Tuple{typeof(ROOT)}, Core.svec())
```
and then just use `ROOTmi` directly in `reset_timings`.